rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

```diff
@@ -0,0 +1,265 @@
+# RFC 6330 - 2021-04-07 - Converting one log event into multiple log events
+
+This RFC describes an approach for transforming one input log event into multiple log events.
+
+## Scope
+
+In:
+
+- Turning a log event into multiple log events
+
+Out:
+
+- Turning a log event into multiple metrics
+- Turning a metric into multiple metrics
+- Turning a metric into multiple logs
+
+## Motivation
+
+Users have asked [1][] [2][] to be able to transform an incoming event into multiple events. This is useful, for example, when parsing an incoming JSON payload that contains an array for which you'd like to publish an event for each element.
+
+Currently, users are restricted to using the Lua or WASM transform for this, [introducing a substantial bottleneck](https://user-images.githubusercontent.com/316880/105531520-d3643200-5cbf-11eb-8b35-fe1c99e5c254.png).
+
+## Internal Proposal
+
+The proposal is to:
+
+1. Extend `remap` so that if `.` is an array at the end, it will emit one event for each element in that array
+2. Add an `unnest` VRL function that will transform an object into an array of objects using a specified field on the input object
+3. Add an `unnest` transform that can be used for simple cases (could just "expand" into a `remap` transform)
+
+Example input:
+
+```json
+{ "host": "localhost", "events": [{ "message": "foo" }, { "message": "bar" }] }
+```
+
+Remap transform config:
+
+```toml
+[transforms.remap]
+type = "remap"
+source = """
+. = unnest(., "events")
+"""
+```
+
+Ouput:
+
+```json
+{ "host": "localhost", "message": "foo" }
+{ "host": "localhost", "message": "bar" }
+```
+
+This would also be available as an `unnest` transform for simple cases where the user does not need to do any additional mapping:
+
+```toml
+[transforms.unnest]
+type = "unnest"
+field = "events"
+```
+
+Which would have the same effect. I think the implementation could just take advantage of the "expand" feature of transforms that lets them replace themselves at vector start time with another transform.
+
+Additionally, we will provide `only_fields` and `except_fields` as options on the `unnest` function and transform to allow users to select which fields will be kept. These match similar semantics to the `encoding` options on sinks.
+
+Example input:
+
+```json
+{ "timestamp": "2020-12-09T16:09:53+00:00", "host": "localhost", "events": [{ "message": "foo" }, { "message": "bar" }] }
+```
+
+Remap transform config:
+
+```toml
+[transforms.remap]
+type = "remap"
+source = """
+. = unnest(., "events", only_fields: ["host"])
```

Comment:
This is the first time (I believe) where we use a string to represent a path. Is there a reason why you chose this over a path-type?, e.g.:

```suggestion
. = unnest(., .events, only_fields: [.host])
```

This would bring it more in-line with functions like `del`, which also takes a path, but doesn't expand that path to a value, instead using it to manipulate the target event.

There are edge-cases to consider here (e.g. what happens when you provide an indexed-path), but overall it does seem to me like the better choice.

Reply:
Ah, yes, I think I like that. I had actually forgotten we could just take the path without resolving it.

We could even just do:

```
. = unnest(.events)
```

Where it would just pull `events` up one level since the path sufficient information about the parent.

This does reduce the user's ability to dynamically set the parameter based on some other value though (e.g. using a different value for `only_fields` by setting a variable prior).

I could see users being slightly confused that:

```
foo = [.host]
. = unnest(., .events, only_fields = foo)
```

Doesn't work. `delete` has this same behavior though.

I'll think about this a bit more.

It is now occurring to me that `del` is also a bit special since it modifies the argument by reference. I think that is the only function that does that.

Reply:
We can discuss this more in the PR for `unnest`, but I'm inclined to just stick with strings rather than requiring paths. I think this'll be less surprising to users.

Reply:
> It is now occurring to me that `del` is also a bit special since it modifies the argument by reference. I think that is the only function that does that.

That is correct. This was discussed during implementation. There were a few more that directly modified the target (such as `only_fields`), but that one was removed before the 1.0 release.

Ergonomically, it felt more useful to have `del` behave this way, and return the value of the deleted field, as it made renaming fields easier to do (`.foo = del(.bar)`).

> We can discuss this more in the PR for `unnest`, but I'm inclined to just stick with strings rather than requiring paths. I think this'll be less surprising to users.

I agree, we don't need to decide here, but I don't think we should use paths, or at least if we do we should be consistent and make this part of the specification (https://github.com/timberio/vector/issues/3740).